templates/install_puppet.bat.erb

```diff
@@ -12,11 +12,22 @@ for /f "delims=" %%i in ('puppet config print --section agent environment') do s
 if exist %pid_path% del %pid_path%
 @echo %pid%> %pid_path%
 
+SET /A pid_checks_performed=0
+
 :wait_for_pid
 timeout /t 5 /nobreak > NUL
-FOR /F "tokens=*" %%A IN ('tasklist /FI "PID eq %AGENT_PID%" /NH') DO set _task=%%A
-echo %_task% | findstr "No tasks are running" >nul
-IF NOT %errorlevel% == 0 ( GOTO wait_for_pid )
+wmic path Win32_Process where handle=%AGENT_PID% get handle /format:textvaluelist | findstr /I "Handle=%AGENT_PID%"
+
+SET /A pid_checks_performed+=1
+REM Wait for a max of 120 seconds (or 24 iterations) before aborting the upgrade
+IF %errorlevel% == 0 IF %pid_checks_performed% == 24 (
+  REM Adding CustomSource allows EventCreate to work properly
+  reg add HKLM\SYSTEM\CurrentControlSet\services\eventlog\Application\Puppet /v CustomSource /t REG_DWORD /d 1 /f
+  EVENTCREATE /T ERROR /L APPLICATION /SO Puppet /ID 101 /D "Puppet agent upgrade failed while waiting for Puppet process [ID: %AGENT_PID%] to exit."
```

Comment:
Unfortunately, in the event log, this can lead to the following message (until we define 101):

```
The description for Event ID 101 from source Puppet cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.

If the event originated on another computer, the display information had to be saved with the event.

The following information was included with the event: 

Puppet agent upgrade failed while waiting for process [4] to exit.

the message resource is present but the message is not found in the string/message table
```